### ISSUE TITLE
client-go credential plugin is slow compared to running as a command

### ISSUE THREAD

**Issue by aperullo** [int128/kubelogin#122]
Using kubelogin in client-go mode is very slow when compared to simply using the `kubelogin` command to generate the kubeconfig.

Consider 2 users:

```
- name: grunt
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - oidc-login
      - get-token
      - --oidc-issuer-url=https://<URL>/auth/realms/master
      - --oidc-client-id=kubernetes
      - --oidc-client-secret=<SECRET>
      - --certificate-authority=./kube-ca.pem
      command: kubectl
      env: null
- name: grunt-manual
  user:
    auth-provider:
      config:
        client-id: kubernetes
        client-secret: <SECRET>
        extra-scopes: groups
        id-token: <ID_TOKEN>
        idp-certificate-authority: ./kube-ca.pem
        idp-issuer-url: https:/<URL>/auth/realms/master
        refresh-token: <REFRESH_TOKEN>
      name: oidc
```

grunt:
```
> time (kcl get pods)
Open http://localhost:18000 for authentication
[17173:17192:0729/164900.154984:ERROR:browser_process_sub_thread.cc(221)] Waited 3 ms for network service
You got a valid token until 2019-07-29 16:50:00 -0400 EDT
Unable to connect to the server: getting credentials: decoding stdout: couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string "json:\"apiVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
( kubectl get pods; )  0.12s user 0.05s system 5% cpu 3.297 total

> time (kcl get pods)
No resources found.
( kubectl get pods; )  0.14s user 0.03s system 7% cpu 2.173 total

...after token expires and refresh token is needed

> time (kcl get pods)
Open http://localhost:18000 for authentication
[20058:20085:0729/165124.847344:ERROR:browser_process_sub_thread.cc(221)] Waited 9 ms for network service
You got a valid token until 2019-07-29 16:52:25 -0400 EDT
Unable to connect to the server: getting credentials: decoding stdout: couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string "json:\"apiVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
( kubectl get pods; )  0.18s user 0.01s system 6% cpu 2.999 total
```

grunt-manual:
```
> kubelogin --user -grunt-manual --certificate-authority kube-ca.pem
Open http://localhost:18000 for authentication
[14214:14242:0729/164628.365946:ERROR:browser_process_sub_thread.cc(221)] Waited 3 ms for network service
Opening in existing browser session.
You got a valid token until 2019-07-29 16:47:29 -0400 EDT

> time (kcl get pods)
No resources found.
( kubectl get pods; )  0.06s user 0.02s system 9% cpu 0.828 total

...after token expires and refresh token is needed

> time (kcl get pods)
No resources found.
( kubectl get pods; )  0.12s user 0.02s system 12% cpu 1.047 total
```

## TL;DR
Time to issue a command.

| Command | Time without refresh (s) | Time with refresh (s) |
| --- | --- | --- |
| grunt | 2.173s | 2.999s |
| grunt-manual | 0.828s | 1.047s |
| **how much slower?** | **162%** | **186%** |

**Comment by int128** [comment]
Would you show detail logs by `-v` option? I think some retry may cause this issue.

**Comment by aperullo** [comment]
Sure, here is log level `-v1`.

```
10:06:20.455025 get_token.go:30: WARNING: log may contain your secrets such as token or password
10:06:20.455150 tls.go:31: Loading the certificate ./kube-ca.pem
10:06:21.153226 auth.go:58: Verifying the existing token
10:06:21.259044 auth.go:73: You have an expired token at 2019-07-30 09:58:49 -0400 EDT
10:06:21.259088 auth.go:77: Refreshing the token
10:06:21.777614 auth.go:89: Could not refresh the token: could not refresh the token: Post https://<URL>/auth/realms/master/protocol/openid-connect/token: x509: certificate signed by unknown authority
10:06:21.777678 auth.go:93: Performing the authentication code flow
Open http://localhost:8000 for authentication
[26735:26754:0730/100622.615542:ERROR:browser_process_sub_thread.cc(221)] Waited 3 ms for network service
10:06:23.493152 get_token.go:57: ID token has the claim: iss=https://<URL>/auth/realms/master
10:06:23.493170 get_token.go:57: ID token has the claim: auth_time=1564494293.000000
10:06:23.493174 get_token.go:57: ID token has the claim: nonce=<NONCE>
10:06:23.493180 get_token.go:57: ID token has the claim: azp=kubernetes
10:06:23.493183 get_token.go:57: ID token has the claim: iat=1564495583.000000
10:06:23.493186 get_token.go:57: ID token has the claim: typ=ID
10:06:23.493189 get_token.go:57: ID token has the claim: exp=1564495643.000000
10:06:23.493191 get_token.go:57: ID token has the claim: nbf=0.000000
10:06:23.493195 get_token.go:57: ID token has the claim: session_state=<STATE>
10:06:23.493198 get_token.go:57: ID token has the claim: aud=kubernetes
10:06:23.493201 get_token.go:57: ID token has the claim: jti=<JTI>
10:06:23.493204 get_token.go:57: ID token has the claim: email_verified=%!s(bool=false)
10:06:23.493207 get_token.go:57: ID token has the claim: groups=[/foo-admin]
10:06:23.493210 get_token.go:57: ID token has the claim: preferred_username=foo
10:06:23.493213 get_token.go:57: ID token has the claim: sub=<SUB>
10:06:23.493216 get_token.go:57: ID token has the claim: acr=0
You got a valid token until 2019-07-30 10:07:23 -0400 EDT
10:06:23.548367 get_token.go:30: WARNING: log may contain your secrets such as token or password
10:06:23.548490 tls.go:31: Loading the certificate ./kube-ca.pem
10:06:24.228210 auth.go:58: Verifying the existing token
10:06:24.335867 auth.go:64: You already have a valid token
10:06:24.335913 get_token.go:57: ID token has the claim: aud=kubernetes
10:06:24.335928 get_token.go:57: ID token has the claim: iss=https://<URL>/auth/realms/master
10:06:24.335939 get_token.go:57: ID token has the claim: sub=<SUB>
10:06:24.335951 get_token.go:57: ID token has the claim: nonce=<NONCE>
10:06:24.335962 get_token.go:57: ID token has the claim: auth_time=1564494293.000000
10:06:24.335981 get_token.go:57: ID token has the claim: nbf=0.000000
10:06:24.335997 get_token.go:57: ID token has the claim: groups=[/foo-admin]
10:06:24.336010 get_token.go:57: ID token has the claim: acr=0
10:06:24.336024 get_token.go:57: ID token has the claim: preferred_username=foo
10:06:24.336043 get_token.go:57: ID token has the claim: typ=ID
10:06:24.336054 get_token.go:57: ID token has the claim: email_verified=%!s(bool=false)
10:06:24.336068 get_token.go:57: ID token has the claim: exp=1564495643.000000
10:06:24.336087 get_token.go:57: ID token has the claim: iat=1564495583.000000
10:06:24.336102 get_token.go:57: ID token has the claim: azp=kubernetes
10:06:24.336124 get_token.go:57: ID token has the claim: jti=<JTI>
10:06:24.336207 get_token.go:57: ID token has the claim: session_state=<STATE>
No resources found.
```

It seems to be unable to refresh the token despite using the same ca-cert in the next step. Then it additionally appears to get the new token, then immediately try to verify the new one before executing the command.

**Comment by int128** [comment]
This is a bug of refreshing token. It needs to wrap the context and use the `http.Client` on refreshing token as well.

https://github.com/int128/kubelogin/blob/79d8056c351333d96e4c77a4e96ee5d768000e71/adaptors/oidc/oidc.go#L74-L78

https://github.com/int128/kubelogin/blob/79d8056c351333d96e4c77a4e96ee5d768000e71/adaptors/oidc/oidc.go#L180-L181

I will release the fix.

> Unrelated but why sometimes does it open http://localhost:18000 and other times it opens http://localhost:8000?

Kubelogin will use port 18000 if port 8000 is not available on your machine.

**Comment by int128** [comment]
Would you check the new release https://github.com/int128/kubelogin/releases/tag/v1.14.1 ? It should fix #121 and #122. Thanks.

**Comment by aperullo** [comment]
After the new version: 

| Command | Time without refresh (s) | Time with refresh (s) |
| --- | --- | --- |
| grunt | 1.79s | 2.71s |
| grunt-manual | 0.740s | 2.550s |

**(Averaged over 10 trials per command)**

It's gotten better, but its still noticeably slower. I can post more logs if that would help.

**Comment by int128** [comment]
This is because kubelogin (grunt) always verifies the cached token and it needs some requests to the provider. On the other hand the auth-provider (grunt-manual) does not verify the token.

It may be possible to omit verifying the cached token. I will try it.

**Comment by aperullo** [comment]
What does it verify about the token? I'll be happy to try it once you do

**Comment by int128** [comment]
It verifies issuer, audience, expiration and signature of the token. It is safe to omit the verification because the token has been verified before caching.

**Comment by int128** [comment]
Would you check the new release? Thanks.

**Comment by aperullo** [comment]
It definitely feels faster. Thank you! 

Time without refresh (s)
`0.08s user 0.02s system 20% cpu 0.479 total`
Time with refresh (s) |
'0.07s user 0.03s system 8% cpu 1.255 total`
